### PR TITLE
[Merged by Bors] - feat(CategoryTheory): characterise locally surjective morphisms in categories of coherent/regular/extensive sheaves 

### PR DESCRIPTION
**Pull request by dagurtomas**
This PR provides simpler characterisations of the property of morphisms of coherent/regular/extensive sheaves of being locally surjective.

In particular, it suffices to check the local surjectivity condition on effective epimorphisms for coherent and regular sheaves, and for extensive sheaves the property is equivalent to being surjective on objects.

---
This is used in #13478 to characterise epimorphisms of condensed sets and modules.

- [x] depends on: #13897 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib.lean [hunk 1]
```diff
@@ -1800,6 +1802,7 @@ import Mathlib.Computability.TMToPartrec
 import Mathlib.Computability.TuringMachine
 import Mathlib.Condensed.Basic
 import Mathlib.Condensed.Discrete
+import Mathlib.Condensed.Epi
```
Comment (dagurtomas) [inline Mathlib.lean:1805]:
```suggestion
```

File: Mathlib/CategoryTheory/Sites/Coherent/ExtensiveTopology.lean [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    | of X S hS =>
+      obtain ⟨α, _, Y, π, h, h'⟩ := hS
+      refine ⟨α, inferInstance, Y, π, ?_, fun a ↦ ?_⟩
+      · have : IsIso (Sigma.desc (Cofan.mk X π).inj) := by simpa using h'
+        exact ⟨Cofan.isColimitOfIsIsoSigmaDesc (Cofan.mk X π)⟩
+      · obtain ⟨rfl, _⟩ := h
+        exact ⟨Y a, 𝟙 Y a, π a, Presieve.ofArrows.mk a, by simp⟩
+    | top X =>
+      refine ⟨Unit, inferInstance, fun _ => X, fun _ => (𝟙 X), ⟨?_⟩, by simp⟩
+      have : IsIso (Sigma.desc (Cofan.mk X fun _ ↦ 𝟙 X).inj) := by
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Sites/Coherent/ExtensiveTopology.lean:42]:
```suggestion
      have : IsIso (Sigma.desc (Cofan.mk X fun (_ : Unit) ↦ 𝟙 X).inj) := by
```

File: Mathlib/CategoryTheory/Sites/Coherent/LocallySurjective.lean [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+universe v w
+
+open CategoryTheory Sheaf Limits Opposite
+
+attribute [local instance] ConcreteCategory.hasCoeToSort ConcreteCategory.instFunLike
+
+namespace CategoryTheory
+
+variable {C : Type*} (D : Type*) [Category C] [Category.{v} D] [ConcreteCategory.{v} D]
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Sites/Coherent/LocallySurjective.lean:38]:
```suggestion
variable {C : Type*} (D : Type*) [Category C] [Category.{v} D] [ConcreteCategory.{w} D]
```

File: Mathlib/CategoryTheory/Sites/LocallySurjective.lean [hunk 4]
```diff
@@ -182,6 +182,11 @@ lemma comp_isLocallySurjective_iff
     IsLocallySurjective J (f₁ ≫ f₂) ↔ IsLocallySurjective J f₂ :=
   isLocallySurjective_iff_of_fac J rfl
 
+variable {J} in
+lemma isLocallySurjectiveMono {K : GrothendieckTopology C} (hJK : J ≤ K) {F G : Cᵒᵖ ⥤ A}
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Sites/LocallySurjective.lean:186]:
```suggestion
lemma isLocallySurjective_of_le {K : GrothendieckTopology C} (hJK : J ≤ K) {F G : Cᵒᵖ ⥤ A}
```

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**